### ISSUE TITLE
Private types

### ISSUE THREAD

**Issue by saleynik** [ejpcmac/typed_struct#27]
It doesn't seem to be possible to define structs with private types. E.g.:
```elixir
defstruct [
  a: nil,
  b: nil
]
@typep  t :: %__MODULE__{
  a: integer,
  b: binary
}
``` 
It would be great if `typedstruct` had a `visibility` or `accessor` attribute to control `private|public` visibility.

**Comment by ejpcmac** [comment]
Indeed, this is a miss. Not sure if there are much use cases for it, since the struct itself cannot be made private, but there is no reason for this feature not to be present.

We already have `opaque: true`, which we could replace in a later version by `visibility:` which could be `:public` (default), `:private` or `:opaque`.

**Comment by ejpcmac** [comment]
*This is a design comment*

## Concept

Let’s add a `visibility` option to `typedstruct` that can be `:public` (default), `:private` or `:opaque`.

## Requirements

* [ ] `typedstruct/2` shall accept a `visibility` option whose values can be `:public`, `:private` or `:opaque`
* [ ] The `visibility` parameter shall be `:public` when not specified
* [ ] `visibility: :public` shall generate a public type
* [ ] `visibility: :private` shall generate a private type
* [ ] `visibility: :opaque` shall generate an opaque type
* [ ] `typedstruct/2` shall still accept the `opaque` option for backward compatibility
* [ ] `opaque: :true` shall have the same effect as `visibility: :opaque`
* [ ] `typedstruct/2` shall not accept both `visibility` and `opaque` options at the same time
* [ ] `typedstruct/2` shall emit a deprecation warning when the `opaque` option is used
